src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1239 @@
+        error_deserializer: Callable[[Any], RiverError],
+        span: Span,
+    ) -> A:
+        """Sends an upload request to the server.
+
+        Expects the input and output be messages that will be msgpacked.
+        """
+        stream_id = nanoid.generate()
+        await self._enqueue_message(
+            stream_id=stream_id,
+            control_flags=STREAM_OPEN_BIT,
+            service_name=service_name,
+            procedure_name=procedure_name,
+            payload=init_serializer(init),
+            span=span,
+        )
+
+        async with self._with_stream(stream_id, 1) as (error_channel, output):
+            try:
+                # If this request is not closed and the session is killed, we should
+                # throw exception here
+                async for item in request:
+                    # Block for backpressure and emission errors from the ws
+                    if err := await error_channel.get():
+                        raise err
+
+                    try:
+                        payload = request_serializer(item)
+                    except Exception as e:
+                        await self._send_cancel_stream(
+                            stream_id=stream_id,
+                            message="Request serialization error",
+                            span=span,
+                        )
+                        raise RiverServiceException(
+                            ERROR_CODE_STREAM_CLOSED,
+                            str(e),
+                            service_name,
+                            procedure_name,
+                        ) from e
+                    await self._enqueue_message(
+                        stream_id=stream_id,
+                        service_name=service_name,
+                        procedure_name=procedure_name,
+                        control_flags=0,
+                        payload=payload,
+                        span=span,
+                    )
+            except Exception as e:
+                # If we get any exception other than WebsocketClosedException,
+                # cancel the stream.
+                await self._send_cancel_stream(
+                    stream_id=stream_id,
+                    message="Unspecified error",
+                    span=span,
+                )
+                raise RiverServiceException(
+                    ERROR_CODE_STREAM_CLOSED, str(e), service_name, procedure_name
+                ) from e
+            await self._send_close_stream(
+                stream_id=stream_id,
+                span=span,
+            )
+
+            try:
+                result = await output.get()
+            except ChannelClosed as e:
+                raise RiverServiceException(
+                    ERROR_CODE_STREAM_CLOSED,
+                    "Stream closed before response",
+                    service_name,
+                    procedure_name,
+                ) from e
+            except Exception as e:
```

Comment:
when does this exception case happen?

Reply:
Really seems to only be if the `Session` is closed. I'll remove since the underlying exception should be sufficient.